### ISSUE TITLE
Problems with L1 Object Map becoming extremely large

### ISSUE THREAD

**Issue by mulhearn** [cms-l1t-offline/cmssw#306]
It seems that in some events the GlobalObjectMapRecord is growing to extremely large size:

https://hypernews.cern.ch/HyperNews/CMS/get/edmFramework/3650.html

we need to sort out why this is happening and fix it.

**Comment by mulhearn** [comment]
To see the issue:

Cherry-pick these commits (into e.g. CMSSW_8_0_7) to setup the tests:

```
011a2499f879eb6abdcb0d5a4d1468a4cbd5bfcb
7bd8aca8333ecf9e885c90de7b57a736829017fa
```

Generate a test script from cmsDriver like this:
`
cmsDriver.py L1TEST --python_filename=l1t_test.py --conditions=80X_dataRun2_v13 -s RAW2DIGI -n 100 --era=Run2_2016 --data --filein=/store/data/Run2016A/ZeroBias1/RAW/v1/000/271/336/00000/00963A5A-BF0A-E611-A657-02163E0141FB.root --no_output --no_exec --customise=L1Trigger/Configuration/customiseHlt.L1TTestHLTSeeds --customise=L1Trigger/Configuration/customiseUtils.L1TStage2DigisSummary
`
Add the example problematic import from the hypernews thread above:

```
process.source = cms.Source("NewEventStreamFileReader",
                            skipEvents = cms.untracked.uint32(1810),
                            fileNames = cms.untracked.vstring('/store/t0streamer/Data/PhysicsHadronsTaus/000/272/798/run272798_ls1472_streamPhysicsHadronsTaus_StorageManager.dat') 
```

**Comment by mulhearn** [comment]
What I see is that this event has a large number of jets:

```
INFO:  L1T found tau collection.
bx:  0  et:  79  eta:  -1.392  phi:  1.914
bx:  0  et:  66.5  eta:  1.044  phi:  -1.32419
bx:  0  et:  53.5  eta:  0  phi:  0.174
bx:  0  et:  39  eta:  0.783  phi:  -2.10719
bx:  0  et:  38  eta:  -0.087  phi:  2.088
bx:  0  et:  36  eta:  0.696  phi:  -2.54219
bx:  0  et:  34.5  eta:  -0.435  phi:  -0.454185
bx:  0  et:  29  eta:  -0.957  phi:  1.479
bx:  0  et:  29  eta:  -0.783  phi:  -0.802185
bx:  0  et:  24.5  eta:  0  phi:  -2.97719
bx:  0  et:  23.5  eta:  2.001  phi:  -2.32469
bx:  0  et:  22  eta:  0.174  phi:  -2.97719
INFO:  L1T found jet collection.
bx:  0  et:  122.5  eta:  -1.4355  phi:  2.001
bx:  0  et:  104  eta:  0.0435  phi:  0.174
bx:  0  et:  98  eta:  1.0005  phi:  -1.32419
bx:  0  et:  81.5  eta:  -1.0005  phi:  1.479
bx:  0  et:  68.5  eta:  -0.0435  phi:  2.088
bx:  0  et:  67.5  eta:  0.8265  phi:  -2.02019
bx:  0  et:  64.5  eta:  -0.0435  phi:  0.696
bx:  0  et:  62  eta:  -0.8265  phi:  -0.715185
bx:  0  et:  58.5  eta:  0.7395  phi:  -2.54219
bx:  0  et:  53  eta:  -0.3915  phi:  -3.06419
bx:  0  et:  44  eta:  2.001  phi:  -2.28119
```

and so perhaps not suprisingly, a large number of combinatorics:

```
algoName:  L1_QuadJetC36_TauJet52
combination vector size:  2
 - size:  7920
 - max size of nested vector:  2880
 - size:  3
 - max size of nested vector:  1
```

There are two types of objects, jets and taus, so there are two elements in the combination vector.
There are 11 central jets above 36, so `11*10*9*8 = 7920` jet combinations.
There are 3 taus above 52, hence the size 3.

That explains size: 7920 and size: 3.

Need to refresh my memory to understand where the 2880 comes from.

**Comment by fwyzard** [comment]
>  There are 11 central jets above 36, so 11 \* 10 \* 9 \* 8 = 7920 jet combinations.

Shouldn't it be (11 \* 10 \* 9 \* 8) / (4 \* 3 \* 2 \* 1) ?
We do not really care about the order of the four jets...

**Comment by mulhearn** [comment]
Actually, do you even need more than 11?

**Comment by mulhearn** [comment]
It looks like in HLT seeding we just, you know, care about which N (e.g. 11) objects possibly seeded a certain M-object (e.g. 4) condition, but we don't actually anywhere check that the corresponding M-objects that pass HLT requirement were one of the actual combinations that passed L1 (e.g. 7920 or 7920/24).

**Comment by mulhearn** [comment]
Sleeping is helpful...  

(1) I'm thinking that 2880 is likely a bug.  Shouldn't that be always 4 for a 4-object condition?  Our HLTSeedFilter removes duplicates, so an array not being reset would go unnoticed downstream.  And this might only get notably large when the outer loop gets large (e.g. (N!)^2).

(2) The choice of which enumerations to explicitly enumerate, which we inherited from legacy, also looks to me now as a sub-obtimal choice.  What HLT needs is much less (e.g. 1 list of 11 items compared to 7920 lists of 4 items) than we provide, but yet what we provide is anyway still not enough to handle every conceivable downstream combinatorics issues.  So it is more than sufficient but not complete, which seems a poor choice to me.  Even if we fix the bug, I think we should switch to a convention where we report 1 list of 11 items (in this example), i.e., the minimum needed by HLT.  This wouldn't require any code change except on uGT side, DataFormat and HLTL1TSeed filter would be unchanged.

**Comment by mulhearn** [comment]
Thanks to Giovanni for finding the culprit, indeed it was as I suspected.  With his proposed fix extended to all calo conditions I find:

```
algoName:  L1_QuadJetC36_TauJet52
combination vector size:  2
 - size:  7920
 - max size of nested vector:  4
 - size:  3
 - max size of nested vector:  1

```

as expected for present implementation.

**Comment by mulhearn** [comment]
This is now fixed with PRs cms-sw#14433 and cms-sw#14431 (merged).

**Comment by mulhearn** [comment]
These evidently did the trick.
